dali/pipeline/operators/reader/coco_reader_op.h

```diff
@@ -25,42 +25,82 @@
 #include <memory>
 
 #include "dali/pipeline/operators/reader/reader_op.h"
-#include "dali/pipeline/operators/reader/loader/file_loader.h"
 #include "dali/pipeline/operators/reader/loader/coco_loader.h"
-#include "dali/pipeline/operators/reader/parser/coco_parser.h"
 
 namespace dali {
-
 class COCOReader : public DataReader<CPUBackend, ImageLabelWrapper> {
  public:
-  explicit COCOReader(const OpSpec& spec)
-  : DataReader<CPUBackend, ImageLabelWrapper>(spec) {
+  explicit COCOReader(const OpSpec& spec):
+    DataReader<CPUBackend, ImageLabelWrapper>(spec),
+    save_img_ids_(spec.GetArgument<bool>("save_img_ids")) {
+    ValidateOptions(spec);
     bool shuffle_after_epoch = spec.GetArgument<bool>("shuffle_after_epoch");
-
-    DALI_ENFORCE(!skip_cached_images_,
-      "COCOReader doesn't support `skip_cached_images` option");
-
-    if (spec.HasArgument("file_list"))
-      loader_ = InitLoader<FileLoader>(
-        spec,
-        std::vector<std::pair<string, int>>(),
-        shuffle_after_epoch);
-    else
-      loader_ = InitLoader<CocoLoader>(
-        spec,
-        annotations_multimap_,
-        shuffle_after_epoch);
-    parser_.reset(new COCOParser(spec, annotations_multimap_));
+    loader_ = InitLoader<CocoLoader>(
+      spec,
+      offsets_,
+      boxes_,
+      labels_,
+      counts_,
+      save_img_ids_,
+      original_ids_,
+      shuffle_after_epoch);
   }
 
   void RunImpl(SampleWorkspace* ws, const int i) override {
-    parser_->Parse(GetSample(ws->data_idx()), ws);
+    const ImageLabelWrapper& image_label = GetSample(ws->data_idx());
+
+    Index image_size = image_label.image.size();
+    auto &image_output = ws->Output<CPUBackend>(0);
+    int image_id = image_label.label;
+
+    image_output.Resize({image_size});
+    image_output.mutable_data<uint8_t>();
+    std::memcpy(
+      image_output.raw_mutable_data(),
+      image_label.image.raw_data(),
+      image_size);
+    image_output.SetSourceInfo(image_label.image.GetSourceInfo());
+
+    auto &boxes_output = ws->Output<CPUBackend>(1);
+    boxes_output.Resize({counts_[image_id], 4});
```

Comment:
Should we define somewhere that box consists of `4` values or it's just nitpicking?

Reply:
This code is connected to `COCO` format and there box has always 4 values. For me literal 4 is more readable here.

Reply:
Sure

Reply:
You can use some constant. This `4` reappears in many places in the code.